delta-lake/delta-spark321db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuDoAutoCompaction.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * This file was derived from DoAutoCompaction.scala
+ * from https://github.com/delta-io/delta/pull/1156
+ * in the Delta Lake project at https://github.com/delta-io/delta.
+ *
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.databricks.sql.transaction.tahoe.rapids
+
+import com.databricks.sql.transaction.tahoe._
+import com.databricks.sql.transaction.tahoe.actions.Action
+import com.databricks.sql.transaction.tahoe.hooks.PostCommitHook
+import com.databricks.sql.transaction.tahoe.metering.DeltaLogging
+
+import org.apache.spark.sql.SparkSession
+
+object GpuDoAutoCompaction extends PostCommitHook
+    with DeltaLogging
+    with Serializable {
+  override val name: String = "GpuDoAutoCompaction"
+
+  override def run(spark: SparkSession,
+                   txn: OptimisticTransactionImpl,
+                   committedActions: Seq[Action]): Unit = {
+    val gpuTxn = txn.asInstanceOf[GpuOptimisticTransaction]
+    val newTxn = new GpuDeltaLog(gpuTxn.deltaLog, gpuTxn.rapidsConf).startTransaction()
+    new GpuOptimizeExecutor(spark, newTxn, Seq.empty, Seq.empty, committedActions).optimize()
```

Comment:
I had trouble getting this working with the existing `OptimizeExec`. I'll give this another go.

Reply:
Closing the loop here:
I tried using the `AutoCompact` pre-commit hook with `GpuOptimisticTransaction`, to achieve the same thing with less code. This doesn't work, because it appears that `AutoCompact` does not call `OptimisticTransaction.writeFiles` at all. It uses `DeltaLog` to write and commit its own transactions. `GpuDeltaLog` can't extend `DeltaLog` because `DeltaLog` is effectively `final`.

I'm afraid this can't be resolved.